// @graphql-tools/stitch/esm/createMergedTypeResolver.js
import { getNamedType, GraphQLList } from 'graphql';
import { batchDelegateToSchema } from '@graphql-tools/batch-delegate';
import { delegateToSchema, } from '@graphql-tools/delegate';
export function createMergedTypeResolver(mergedTypeResolverOptions, mergedType) {
    const { fieldName, argsFromKeys, valuesFromResults, args } = mergedTypeResolverOptions;
    if (argsFromKeys != null) {
        return function mergedBatchedTypeResolver(_originalResult, context, info, subschema, selectionSet, key, type = mergedType || getNamedType(info.returnType)) {
            return batchDelegateToSchema({
                schema: subschema,
                operation: 'query',
                fieldName,
                returnType: new GraphQLList(type),
                key,
                argsFromKeys,
                valuesFromResults,
                selectionSet,
                context,
                info,
                skipTypeMerging: true,
                dataLoaderOptions: mergedTypeResolverOptions.dataLoaderOptions,
            });
        };
    }
    if (args != null) {
        return function mergedTypeResolver(originalResult, context, info, subschema, selectionSet, _key, type = mergedType || getNamedType(info.returnType)) {
            return delegateToSchema({
                schema: subschema,
                operation: 'query',
                fieldName,
                returnType: type,
                args: args(originalResult),
                selectionSet,
                context,
                info,
                skipTypeMerging: true,
            });
        };
    }
    return undefined;
}
