// @graphprotocol/client-block-tracking/cjs/shared.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMetaSelectionNode = exports.DEFAULT_CONFIG = void 0;
exports.transformExecutionRequest = transformExecutionRequest;
exports.getNewBlockNumberFromExecutionResult = getNewBlockNumberFromExecutionResult;
const utils_1 = require("@graphql-tools/utils");
const graphql_1 = require("graphql");
exports.DEFAULT_CONFIG = {
    ignoreOperationNames: [],
    ignoreFieldNames: [],
    metaTypeName: '_Meta_',
    blockFieldName: 'block',
    blockNumberFieldName: 'number',
    metaRootFieldName: '_meta',
    blockArgumentName: 'block',
    minBlockArgumentName: 'number_gte',
};
exports.createMetaSelectionNode = (0, utils_1.memoize1)(function createMetaSelectionNode(config) {
    return {
        kind: graphql_1.Kind.FIELD,
        name: {
            kind: graphql_1.Kind.NAME,
            value: config.metaRootFieldName,
        },
        selectionSet: {
            kind: graphql_1.Kind.SELECTION_SET,
            selections: [
                {
                    kind: graphql_1.Kind.FIELD,
                    name: {
                        kind: graphql_1.Kind.NAME,
                        value: config.blockFieldName,
                    },
                    selectionSet: {
                        kind: graphql_1.Kind.SELECTION_SET,
                        selections: [
                            {
                                kind: graphql_1.Kind.FIELD,
                                name: {
                                    kind: graphql_1.Kind.NAME,
                                    value: config.blockNumberFieldName,
                                },
                            },
                        ],
                    },
                },
            ],
        },
    };
});
const getQueryFieldNames = (0, utils_1.memoize1)(function getQueryFields(schema) {
    const queryType = schema.getQueryType();
    if (queryType == null) {
        throw new Error(`Make sure you have a query type in this source before applying Block Tracking`);
    }
    return Object.keys(queryType.getFields());
});
const metaFieldAddedByContext = new WeakMap();
function getRequestIdentifier(executionRequest) {
    return executionRequest.context ?? executionRequest.rootValue ?? executionRequest.info?.operation;
}
function transformExecutionRequest(executionRequest, config, transformedSchema, batch = false, minBlock) {
    if (executionRequest.operationName != null && config.ignoreOperationNames.includes(executionRequest.operationName)) {
        return executionRequest;
    }
    const document = (0, graphql_1.visit)(executionRequest.document, {
        Field: {
            leave: (fieldSelectionNode) => {
                if (minBlock != null &&
                    !fieldSelectionNode.name.value.startsWith('_') &&
                    !config.ignoreFieldNames.includes(fieldSelectionNode.name.value) &&
                    getQueryFieldNames(transformedSchema).includes(fieldSelectionNode.name.value)) {
                    const argNodes = fieldSelectionNode.arguments?.filter((argument) => argument.name.value !== config.blockArgumentName) || [];
                    const blockArgument = argNodes.find((argument) => argument.name.value === config.blockArgumentName) || {
                        kind: graphql_1.Kind.ARGUMENT,
                        name: {
                            kind: graphql_1.Kind.NAME,
                            value: config.blockArgumentName,
                        },
                        value: {
                            kind: graphql_1.Kind.OBJECT,
                            fields: [],
                        },
                    };
                    const blockArgumentFields = blockArgument.value.fields;
                    if (!blockArgumentFields.some((field) => field.name.value === config.minBlockArgumentName)) {
                        return {
                            ...fieldSelectionNode,
                            arguments: [
                                ...argNodes,
                                {
                                    ...blockArgument,
                                    value: {
                                        ...blockArgument.value,
                                        fields: [
                                            ...blockArgumentFields,
                                            {
                                                kind: graphql_1.Kind.OBJECT_FIELD,
                                                name: {
                                                    kind: graphql_1.Kind.NAME,
                                                    value: config.minBlockArgumentName,
                                                },
                                                value: {
                                                    kind: graphql_1.Kind.INT,
                                                    value: minBlock.toString(),
                                                },
                                            },
                                        ],
                                    },
                                },
                            ],
                        };
                    }
                }
                return fieldSelectionNode;
            },
        },
        OperationDefinition: {
            leave: (operationNode) => {
                const requestIdentifier = getRequestIdentifier(executionRequest);
                let shouldAddMetaField = true;
                if (batch) {
                    if (requestIdentifier != null) {
                        const isAddedBefore = metaFieldAddedByContext.get(requestIdentifier);
                        if (isAddedBefore != null) {
                            shouldAddMetaField = !isAddedBefore;
                        }
                    }
                }
                if (operationNode.operation === graphql_1.OperationTypeNode.QUERY && shouldAddMetaField) {
                    const metaSelectionNode = (0, exports.createMetaSelectionNode)(config);
                    metaFieldAddedByContext.set(requestIdentifier, true);
                    return {
                        ...operationNode,
                        selectionSet: {
                            ...operationNode.selectionSet,
                            selections: [...operationNode.selectionSet.selections, metaSelectionNode],
                        },
                    };
                }
                return operationNode;
            },
        },
    });
    return {
        ...executionRequest,
        document,
    };
}
function getNewBlockNumberFromExecutionResult(originalResult, config) {
    return originalResult.data?.[config.metaRootFieldName]?.[config.blockFieldName]?.[config.blockNumberFieldName];
}
